### PR TITLE
--add-compiler-defaults option added. When cross compiling this passe…

### LINKED ISSUE #426: builtin parameters of cross-compilers  are not handled
cross compilers may have built-in defines and include paths

These are invisible for the ld-logger and thus not represented in the llvm comilation db.
Therefore the clang analysis frequently fails.

CodeChecker should detect built-in compiler defines and include paths and pass it to clang when running the analysis.

### PR DESCRIPTION
**Pull request by dkrupp**
…s default compiler inlucdes,defines to clang. fixes #426

### CODE REVIEW

File: codechecker_lib/log_parser.py [hunk 1]
```diff
@@ -15,15 +18,80 @@
...
+        out, err = proc.communicate("")
+        for line in out.splitlines(True):
+            LOG.debug("define:"+line)
+            define = line.strip().split(" ")[1:]
+            d = "-D"+define[0] + '=' + '"' + ' '.join(define[1:]) + '"'
+            defines.append(d)
+
+    except OSError as oerr:
+        LOG.error("Cannot find defines:" + oerr.strerror+"\n")
+    return defines
```
Comment (gyorb) [inline codechecker_lib/log_parser.py]: `finally` can be used here too like in the `get_compiler_includes` method before.

File: codechecker_lib/log_parser.py [hunk 2]
```diff
@@ -15,15 +18,80 @@
...
+
+
+# -----------------------------------------------------------------------------
+def parse_compile_commands_json(logfile, add_compiler_defaults=False):
     import json
+    LOG.debug_analyzer('parse_compile_commands_json: ' +
```
Comment (gyorb) [inline codechecker_lib/log_parser.py]: This can be LOG.debug message.

**Comment by gyorb** [comment]
Please update your branch, there is a fix for Travis OSX build.

File: codechecker_lib/log_parser.py [hunk 3]
```diff
@@ -15,15 +18,80 @@
 
 
 # -----------------------------------------------------------------------------
-def parse_compile_commands_json(logfile):
+def get_compiler_includes(compiler):
+    """
+    Returns a list of default includes of the given compiler.
+    """
+    LOG.debug_analyzer('getting include paths for  ' + compiler)
```
Comment (dkrupp) [inline codechecker_lib/log_parser.py]: fixed

File: codechecker_lib/log_parser.py [hunk 4]
```diff
@@ -15,15 +18,80 @@
...
+    include_paths = []
+    try:
+        proc = subprocess.Popen(shlex.split(cmd),
+                                stdin=subprocess.PIPE,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+
+        out, err = proc.communicate("")
+
+        print_line = False
```
Comment (dkrupp) [inline codechecker_lib/log_parser.py]: changed to:do_append

File: codechecker_lib/log_parser.py [hunk 5]
```diff
@@ -15,15 +18,80 @@
...
+        for line in err.splitlines(True):
+            if line.startswith(end_mark):
```
Comment (dkrupp) [inline codechecker_lib/log_parser.py]: fixed

File: codechecker_lib/log_parser.py [hunk 6]
```diff
@@ -15,15 +18,80 @@
...
+
+
+# -----------------------------------------------------------------------------
+def get_compiler_defines(compiler):
+    """
+    Returns a list of default defines of the given compiler.
+    """
+    cmd = compiler + " -dM -E -"
+    defines = []
+    FNULL = open(os.devnull, 'r')
```
Comment (dkrupp) [inline codechecker_lib/log_parser.py]: fixed

File: codechecker_lib/log_parser.py (hunk 1 above)
Comment (dkrupp) [inline codechecker_lib/log_parser.py]: finally removed

File: codechecker_lib/log_parser.py (hunk 2 above)
Comment (dkrupp) [inline codechecker_lib/log_parser.py]: fixed

**Pull request merged**